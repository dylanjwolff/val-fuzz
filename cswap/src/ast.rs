use crate::Timer;
use std::cell::RefCell;
use std::fmt;
use std::fs;
use std::path::Path;
use std::rc::Rc;
use std::time::Duration;
#[macro_use]
use serde::{Serialize, Deserialize};

pub type ScriptRc = Rc<RefCell<Script>>;
pub type CommandRc = Rc<RefCell<Command>>;
pub type LogicRc = Rc<RefCell<Logic>>;
pub type SExpRc = Rc<RefCell<SExp>>;
pub type SExpBoxRc = Rc<RefCell<Box<SExp>>>;
pub type SortRc = Rc<RefCell<Sort>>;
pub type SymbolRc = Rc<RefCell<Symbol>>;
pub type ConstantRc = Rc<RefCell<Constant>>;
pub type BoolOpRc = Rc<RefCell<BoolOp>>;

macro_rules! rccell {
    ($x:expr) => {{
        std::rc::Rc::new(std::cell::RefCell::new($x))
    }};
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Script {
    Commands(Vec<CommandRc>),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Command {
    Logic(LogicRc),
    CheckSat(),
    CheckSatAssuming(SExpRc),
    Assert(SExpRc),
    GetModel(),
    DeclConst(String, SortRc),
    Generic(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Sort {
    UInt(),
    Dec(),
    Str(),
    Bool(),
    BitVec(u32),
    Array(),
    UserDef(String),
    Compound(Vec<SortRc>),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum SExp {
    Compound(Vec<SExpRc>),
    Let(Vec<(SymbolRc, SExpRc)>, SExpBoxRc),
    QForAll(Vec<(SymbolRc, SortRc)>, SExpBoxRc),
    BExp(BoolOpRc, Vec<SExpRc>),
    Constant(ConstantRc),
    Symbol(SymbolRc),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Symbol {
    Var(String), // Currently, it is hard to detect all variables, so some Tokens may also be
    // variables too
    Token(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum BoolOp {
    Equals(),
    And(),
    Or(),
    Xor(),
    Implies(),
    Distinct(),
    Gt(),
    Lt(),
    Gte(),
    Lte(),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Constant {
    UInt(String),
    Dec(String),
    Hex(String),
    Bin(Vec<char>),
    Str(String),
    Bool(bool),
}

#[allow(non_camel_case_types)]
#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Logic {
    QF_SLIA(),
    Other(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum AstNode {
    Script(ScriptRc),
    Command(CommandRc),
    Constant(ConstantRc),
    Symbol(SymbolRc),
    SExp(SExpRc),
    Logic(LogicRc),
    BoolOp(BoolOpRc),
    Sort(SortRc),
    Open(),
    Close(),
}

impl Script {
    pub fn to_f(&mut self, filepath: &Path) {
        fs::write(filepath, self.to_string());
    }

    pub fn insert(&mut self, i: usize, cmd: Command) {
        let Script::Commands(cmds) = self;
        if i > cmds.len() {
            cmds.push(rccell!(cmd));
        } else {
            cmds.insert(i, rccell!(cmd));
        }
    }

    pub fn replace(&mut self, i: usize, cmd: Command) {
        let Script::Commands(cmds) = self;
        cmds[i] = rccell!(cmd);
    }

    pub fn init(&mut self, i: usize) {
        let Script::Commands(cmds) = self;
        cmds.insert(i, rccell!(Command::Assert(rccell!(SExp::true_sexp()))));
    }

    pub fn index_is_gm(&self, i: usize) -> bool {
        let Script::Commands(cmds) = self;
        if i >= cmds.len() {
            return false;
        }
        match &*cmds[i].borrow() {
            Command::GetModel() => true,
            _ => false,
        }
    }

    pub fn is_unsupported_logic(&self) -> bool {
        let Script::Commands(cmds) = self;
        let maybe_logic = cmds.iter().find(|cmd| cmd.borrow().is_logic());
        match maybe_logic {
            Some(cmd) => match &*cmd.borrow() {
                Command::Logic(l) => match *l.borrow() {
                    Logic::QF_SLIA() => true,
                    _ => false,
                },
                _ => false,
            },
            _ => false,
        }
    }
}

impl fmt::Display for Script {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Script::Commands(cmds) => cmds
                .iter()
                .map(|cmd| write!(f, "{}\n", cmd.borrow()))
                .fold(Ok(()), acc_result),
        }
    }
}

impl Command {
    pub fn is_logic(&self) -> bool {
        match self {
            Command::Logic(_) => true,
            _ => false,
        }
    }

    pub fn is_checksat(&self) -> bool {
        match self {
            Command::CheckSat() | Command::CheckSatAssuming(_) => true,
            _ => false,
        }
    }

    pub fn no_op() -> Self {
        Command::Generic("".to_string())
    }
}

impl fmt::Display for Command {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Command::Logic(l) => write!(f, "(set-logic {})", l.borrow()),
            Command::CheckSat() => write!(f, "(check-sat)"),
            Command::CheckSatAssuming(sexp) => write!(f, "(check-sat-assuming {})", sexp.borrow()),
            Command::GetModel() => write!(f, "(get-model)"),
            Command::DeclConst(v, s) => write!(f, "(declare-const {} {})", v, s.borrow()),
            Command::Generic(s) => write!(f, "{}", s),
            Command::Assert(s) => write!(f, "(assert {})", s.borrow()),
        }
    }
}

impl fmt::Display for Constant {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Constant::UInt(s) => write!(f, "{}", s),
            Constant::Dec(d) => write!(f, "{}", d),
            Constant::Hex(s) => write!(f, "#x{}", s),
            Constant::Str(s) => write!(f, "\"{}\"", s),
            Constant::Bool(b) => write!(f, "{}", b),
            Constant::Bin(bv) => write!(f, "#b{}", bv.into_iter().collect::<String>()),
        }
    }
}

fn acc_result(acc: fmt::Result, curr: fmt::Result) -> fmt::Result {
    match curr {
        Ok(_) => acc,
        Err(e) => Err(e),
    }
}

impl fmt::Display for Sort {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Sort::UInt() => write!(f, "Int"),
            Sort::Dec() => write!(f, "Real"),
            Sort::Bool() => write!(f, "Bool"),
            Sort::Str() => write!(f, "String"),
            Sort::BitVec(len) => write!(f, "(_ BitVec {})", len),
            Sort::Array() => write!(f, "Array"),
            Sort::UserDef(s) => write!(f, "{}", s),
            Sort::Compound(v) => {
                write!(f, "(")?;
                v.iter()
                    .enumerate()
                    .map(|(i, sort)| match i == 0 {
                        true => write!(f, "{}", sort.borrow()),
                        false => write!(f, " {}", sort.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
        }
    }
}

impl SExp {
    pub fn true_sexp() -> SExp {
        SExp::Symbol(rccell!(Symbol::Token("true".to_owned())))
    }

    pub fn false_sexp() -> SExp {
        SExp::Symbol(rccell!(Symbol::Token("false".to_owned())))
    }
}

impl fmt::Display for SExp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SExp::Constant(c) => write!(f, "{}", c.borrow()),
            SExp::Symbol(s) => write!(f, "{}", s.borrow()),
            SExp::Let(vbs, s) => {
                write!(f, "(let (")?;
                vbs.iter()
                    .map(|(v, s)| write!(f, "({} {})", v.borrow(), s.borrow()))
                    .fold(Ok(()), acc_result)?;
                write!(f, " {})", s.borrow())
            }
            SExp::Compound(v) => {
                write!(f, "(")?;
                v.iter()
                    .enumerate()
                    .map(|(i, sexp)| match i == 0 {
                        true => write!(f, "{}", sexp.borrow()),
                        false => write!(f, " {}", sexp.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
            SExp::BExp(o, v) => {
                write!(f, "({} ", o.borrow())?;
                v.iter()
                    .enumerate()
                    .map(|(i, sexp)| match i == 0 {
                        true => write!(f, "{}", sexp.borrow()),
                        false => write!(f, " {}", sexp.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
            SExp::QForAll(v, s) => {
                write!(f, "(forall (")?;
                v.iter()
                    .map(|(va, vl)| write!(f, "({} {})", va.borrow(), vl.borrow()))
                    .fold(Ok(()), acc_result)?;
                write!(f, ") {})", s.borrow())
            }
        }
    }
}

impl fmt::Display for Symbol {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Symbol::Var(s) | Symbol::Token(s) => write!(f, "{}", s),
        }
    }
}

impl fmt::Display for BoolOp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            BoolOp::And() => write!(f, "and"),
            BoolOp::Or() => write!(f, "or"),
            BoolOp::Xor() => write!(f, "xor"),
            BoolOp::Implies() => write!(f, "=>"),
            BoolOp::Distinct() => write!(f, "distinct"),
            BoolOp::Equals() => write!(f, "="),
            BoolOp::Gt() => write!(f, ">"),
            BoolOp::Lt() => write!(f, "<"),
            BoolOp::Gte() => write!(f, ">="),
            BoolOp::Lte() => write!(f, "<="),
        }
    }
}

impl fmt::Display for Logic {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Logic::QF_SLIA() => write!(f, "QF_SLIA"),
            Logic::Other(s) => write!(f, "{}", s),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::*;
    use insta::assert_debug_snapshot;
    use walkdir::WalkDir;

    #[test]
    fn sort_display_snap() {
        let sort = Sort::Compound(vec![rccell!(Sort::UInt()), rccell!(Sort::Bool())]);
        assert_debug_snapshot!(format!("{}", sort));
    }

    #[test]
    fn let_to_string_snap() {
        assert_debug_snapshot!(script(
            "(assert (let ((max (seq.nth A m))(n   (seq.len A))) (ite (or true false))))"
        )
        .unwrap()
        .1
        .to_string());
    }
}
