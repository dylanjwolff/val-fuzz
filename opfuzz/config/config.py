solvers = [
]

crash_msgs = [
    "Exception",
    "lang.AssertionError",
    "lang.Error",
    "runtime error",
    "LEAKED",
    "Leaked",
    "Segmentation fault",
    "segmentation fault",
    "segfault",
    "ASSERTION",
    "Assertion",
    "Fatal failure",
    "Internal error detected",
    "an invalid model was generated",
    "Failed to verify",
    "failed to verify",
    "ERROR: AddressSanitizer:",
    "invalid expression",
    "Aborted"
]

ignore_msgs = [
    "unexpected char",
    "failed to open file",
    "Expected result sat but got unsat",
    "Expected result unsat but got sat",
    "Parse Error",
    "Cannot get model",
    "Symbol 'str.to-re' not declared as a variable",
    "Symbol 'str.to.re' not declared as a variable",
    "Unimplemented code encountered",
]
