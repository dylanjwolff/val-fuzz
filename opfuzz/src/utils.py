import random
import string

def random_string(length=5):
    return ''.join(random.sample(string.ascii_letters + string.digits, length))

def plain(cli):
    plain_cli = ""
    for token in cli.split(" "):
        plain_cli += token.split("/")[-1]
    return escape(plain_cli)

def escape(s):
    s = s.replace(".", "")
    s = s.replace("=", "")
    return s

