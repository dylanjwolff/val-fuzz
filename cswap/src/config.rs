use crate::ast::Script;
use std::fs;
use std::io;
use std::path::Path;
use std::path::PathBuf;

use crate::ast::Sort;

use tempfile::Builder;
use tempfile::TempDir;

use crate::solver::ProfileIndex;
use log::warn;
use serde::{Deserialize, Serialize};
use std::collections::hash_map::DefaultHasher;
use std::collections::HashSet;
use std::hash::Hash;
use std::hash::Hasher;

#[derive(Debug, Clone)]
pub struct Config {
    pub file_provider: FileProvider,
    pub remove_files: bool,
    pub max_iter: u32,
    pub rng_seed: u64,
    pub stack_size: usize,
    pub profiles: HashSet<ProfileIndex>,
    pub mask_size: usize,
    pub monitors_in_final: bool,
    pub enforce_on_resub: bool,
    pub use_bdom_vs: bool,
    pub max_const_relations_to_monitor: u8,
    pub skolemize_universal: bool,
    pub leaf_opt: bool,
    pub cp_og: bool,
}

#[macro_export]
macro_rules! liftio {
    ($x:expr) => {
        $x.map_err(|e| io::Error::new(io::ErrorKind::Other, format!("(Lifted) {:?}", e)))
    };
}

#[macro_export]
macro_rules! liftio_e {
    ($x:expr) => {
        io::Error::new(io::ErrorKind::Other, format!("(Lifted) {:?}", $x))
    };
}

impl Config {
    pub fn default() -> Self {
        Config {
            file_provider: FileProvider::new_tmp(),
            rng_seed: 0,
            max_iter: 20,
            stack_size: 50,
            remove_files: true,
            mask_size: 1,
            profiles: HashSet::new(),
            monitors_in_final: false,
            enforce_on_resub: false,
            use_bdom_vs: false,
            max_const_relations_to_monitor: 0,
            skolemize_universal: false,
            leaf_opt: false,
            cp_og: false,
        }
    }

    pub fn get_specific_seed<T: Hash>(&self, t: &T) -> u64 {
        let mut s = DefaultHasher::new();
        t.hash(&mut s);
        self.rng_seed.hash(&mut s);
        s.finish()
    }

    pub fn to_csv_string(&self) -> String {
        format!(
            "{}, {}, {}, {}, {}, {}",
            self.max_iter,
            self.rng_seed,
            self.stack_size,
            self.mask_size,
            self.monitors_in_final,
            self.enforce_on_resub
        )
    }
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct FileProvider {
    pub basedir: PathBuf,
    pub skeldir: PathBuf,
    pub cholesdir: PathBuf,
    pub mddir: PathBuf,
    pub scratchdir: PathBuf,
    pub bugdir: PathBuf,
}

impl FileProvider {
    pub fn new_tmp() -> FileProvider {
        FileProvider::new_existing(TempDir::new().unwrap().into_path().to_str().unwrap(), true)
    }
    pub fn new_existing(dirname: &str, existing: bool) -> FileProvider {
        let dirpath = Path::new(dirname).to_path_buf();
        let cholesdir = Self::get_subdir(&dirpath, "choles");
        let skeldir = Self::get_subdir(&dirpath, "skel");
        let mddir = Self::get_subdir(&dirpath, "md");
        let scratchdir = Self::get_subdir(&dirpath, "scratch");
        let bugdir = Self::get_subdir(&dirpath, "bugs");
        if !existing {
            fs::create_dir(&dirpath).unwrap();
        }
        fs::create_dir(&skeldir).unwrap();
        fs::create_dir(&cholesdir).unwrap();
        fs::create_dir(&mddir).unwrap();
        fs::create_dir(&scratchdir).unwrap();
        fs::create_dir(&bugdir).unwrap();
        FileProvider {
            basedir: dirpath,
            cholesdir: cholesdir,
            skeldir: skeldir,
            mddir: mddir,
            scratchdir: scratchdir,
            bugdir: bugdir,
        }
    }

    pub fn new(dirname: &str) -> FileProvider {
        FileProvider::new_existing(dirname, false)
    }

    pub fn og_w_monitors<'a>(&self, md: &'a mut Metadata) -> io::Result<PathBuf> {
        let tfile = Builder::new()
            .prefix("ogwms_")
            .rand_bytes(0)
            .suffix(&md.seed_file)
            .tempfile_in(&self.skeldir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        md.og_w_monitors_skel_file = Some(name(&pb));
        Ok(pb)
    }

    pub fn cholesfile<'a>(&self, md: &'a mut Metadata) -> io::Result<PathBuf> {
        let tfile = Builder::new()
            .prefix("choles_")
            .rand_bytes(0)
            .suffix(&md.seed_file)
            .tempfile_in(&self.cholesdir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        md.choles_file = name(&pb);
        Ok(pb)
    }

    pub fn skelfile<'a>(&self, md: &'a mut Metadata) -> io::Result<PathBuf> {
        let tfile = Builder::new()
            .prefix("skel_")
            .rand_bytes(2)
            .suffix(&format!("_{}", md.seed_file))
            .tempfile_in(&self.skeldir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        md.skeleton_file = name(&pb);
        Ok(pb)
    }

    pub fn mdfile<'a>(&self, md: &'a mut Metadata) -> io::Result<PathBuf> {
        let tfile = Builder::new()
            .prefix("md_")
            .rand_bytes(2)
            .suffix(&format!("_{}", md.seed_file))
            .tempfile_in(&self.mddir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        md.metadata_file = name(&pb);
        Ok(pb)
    }

    pub fn iterfile(&self, i: u64, md: &Metadata) -> io::Result<PathBuf> {
        let tfile = Builder::new()
            .prefix("iter_")
            .rand_bytes(10)
            .suffix(&format!("_{}_{}", i, md.seed_file))
            .tempfile_in(&self.scratchdir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        Ok(pb)
    }

    pub fn resubfile(&self, iterfile: &Path, md: &Metadata) -> io::Result<PathBuf> {
        let stem = iterfile
            .file_name()
            .and_then(|s| s.to_str())
            .map(|s| s.to_owned())
            .unwrap_or(md.seed_file.to_owned() + "_unknown_iter");
        let tfile = Builder::new()
            .prefix("resub_")
            .rand_bytes(10)
            .suffix(&format!("_{}", stem))
            .tempfile_in(&self.scratchdir)?;
        let (_, pb) = liftio!(tfile.keep())?;
        Ok(pb)
    }

    pub fn bug_report(&self, buggy_file: &Path, report: &str) {
        let r = Builder::new()
            .prefix("bug_report_")
            .rand_bytes(0)
            .suffix(
                buggy_file
                    .file_stem()
                    .and_then(|n| n.to_str())
                    .unwrap_or("default_name"),
            )
            .tempfile_in(&self.bugdir)
            .and_then(|tf| {
                fs::write(tf.path(), report)?;
                liftio!(tf.keep())
            });
        match r {
            Err(e) => warn!("Error writing bug report for {:?}: {}", buggy_file, e),
            _ => (),
        }
    }

    pub fn cleanup_all(&self) {
        match fs::remove_dir_all(&self.basedir) {
            Err(e) => match e.kind() {
                std::io::ErrorKind::NotFound => (),
                _ => panic!(e),
            },
            _ => (),
        }
    }

    pub fn serialize_skel<'a, 'b>(
        &self,
        script: &'a Script,
        md: &'b mut Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.skelfile(md)?;
        fs::write(&f, script.to_string())?;
        Ok(f)
    }

    pub fn serialize_og_w_m<'a, 'b>(
        &self,
        script: &'a Script,
        md: &'b mut Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.og_w_monitors(md)?;
        fs::write(&f, script.to_string())?;
        Ok(f)
    }

    pub fn serialize_md<'a>(&self, md: &'a mut Metadata) -> Result<PathBuf, io::Error> {
        let f = self.mdfile(md)?;
        let mdstr = serde_lexpr::to_string(&md)?;
        fs::write(&f, mdstr)?;
        Ok(f)
    }

    pub fn serialize_skel_and_md<'a, 'b>(
        &self,
        script: &'a Script,
        md: &'b mut Metadata,
    ) -> Result<(PathBuf, PathBuf), io::Error> {
        let fs = self.serialize_skel(script, md)?;
        let fm = self.serialize_md(md)?;
        Ok((fs, fm))
    }

    pub fn serialize_iterfile_str(
        &self,
        script_str: &str,
        iteration: u64,
        md: &Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.iterfile(iteration, md)?;
        fs::write(&f, script_str)?;
        Ok(f)
    }

    pub fn serialize_iterfile(
        &self,
        script: &Script,
        iteration: u64,
        md: &Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.iterfile(iteration, md)?;
        fs::write(&f, script.to_string())?;
        Ok(f)
    }

    pub fn serialize_resub_str(
        &self,
        script_str: String,
        iterfile: &Path,
        md: &Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.resubfile(iterfile, md)?;
        fs::write(&f, script_str)?;
        Ok(f)
    }

    pub fn serialize_resub(
        &self,
        script: &Script,
        iterfile: &Path,
        md: &Metadata,
    ) -> Result<PathBuf, io::Error> {
        let f = self.resubfile(iterfile, md)?;
        fs::write(&f, script.to_string())?;
        Ok(f)
    }

    fn get_subdir(base: &Path, subname: &str) -> PathBuf {
        let mut subdir = base.to_path_buf();
        subdir.push(subname);
        subdir
    }
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub struct Metadata {
    pub bavns: Vec<(String, Sort)>,
    pub constvns: Vec<(String, Sort)>,
    pub seed_file: String,
    pub skeleton_file: String,
    pub og_w_monitors_skel_file: Option<String>,
    pub choles_file: String,
    pub metadata_file: String,
}
pub fn name(pb: &Path) -> String {
    pb.file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("default_filename")
        .to_owned()
}

impl Metadata {
    pub fn new_empty() -> Self {
        Metadata {
            bavns: vec![],
            constvns: vec![],
            seed_file: "".to_owned(),
            skeleton_file: "".to_owned(),
            choles_file: "".to_owned(),
            metadata_file: "".to_owned(),
            og_w_monitors_skel_file: None,
        }
    }
    pub fn new(seed: &Path) -> Self {
        Metadata {
            bavns: vec![],
            constvns: vec![],
            seed_file: name(seed),
            choles_file: "".to_owned(),
            skeleton_file: "".to_owned(),
            metadata_file: "".to_owned(),
            og_w_monitors_skel_file: None,
        }
    }

    pub fn seed_path(&self, _fp: &FileProvider) -> PathBuf {
        panic!("Unimplemented!");
    }

    pub fn skel_path(&self, fp: &FileProvider) -> PathBuf {
        let mut p = fp.skeldir.clone();
        p.push(&self.skeleton_file);
        p
    }

    pub fn md_path(&self, fp: &FileProvider) -> PathBuf {
        let mut p = fp.mddir.clone();
        p.push(&self.metadata_file);
        p
    }

    pub fn choles_path(&self, fp: &FileProvider) -> PathBuf {
        let mut p = fp.cholesdir.clone();
        p.push(&self.choles_file);
        p
    }

    pub fn ogwms_path(&self, fp: &FileProvider) -> Option<PathBuf> {
        match &self.og_w_monitors_skel_file {
            Some(file) => {
                let mut p = fp.skeldir.clone();
                p.push(file);
                Some(p)
            }
            _ => None,
        }
    }
}
