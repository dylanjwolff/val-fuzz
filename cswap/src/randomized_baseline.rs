use crate::ast::BitVecConst;
use crate::ast::Command;
use crate::ast::Constant;
use crate::ast::FPConst;
use crate::ast::SExp;
use crate::ast::Script;
use crate::ast::Sort;
use crate::config::Config;
use crate::config::Metadata;
use crate::parser::script_from_f;
use crate::solver::profiles_solve;
use crate::utils::RunStats;

use crate::fuzzer::report_any_bugs;
use arbitrary::Arbitrary;
use arbitrary::Unstructured;
use log::warn;
use rand::Rng;
use rand_xoshiro::rand_core::SeedableRng;
use rand_xoshiro::Xoshiro256Plus;
use std::fmt;
use std::fs;
use std::io;
use std::path::Path;

pub fn rand_fuzz_iter_and_solve(
    seed_f: &Path,
    cfg: &Config,
    stats: &mut RunStats,
) -> io::Result<()> {
    let mut md = Metadata::new(&seed_f);
    let mut script = script_from_f(&seed_f)?;
    let mut rng = Xoshiro256Plus::seed_from_u64(1);
    for i in 0..cfg.max_iter as u64 {
        fuzz(&mut script, &mut rng);
        let script_str = script.to_string();
        if !stats.record_sub(&script_str) {
            continue; // Reporting this as an error is probably a bit too noisy
        }
        let iter_f = match cfg.file_provider.serialize_iterfile(&script, i, &mut md) {
            Err(e) => {
                warn!("Serialization error {} for {}", e, md.seed_file);
                continue;
            }
            Ok(iter_f) => iter_f,
        };

        let results = profiles_solve(iter_f.to_str().unwrap_or("defaultname"), &cfg.profiles);
        stats.record_stats_for_sub_results(&results);

        if !report_any_bugs(&iter_f, &results, &cfg.file_provider) {
            if cfg.remove_files {
                fs::remove_file(&iter_f).unwrap_or(());
            }
        }
    }

    Ok(())
}

pub fn fuzz(script: &mut Script, rng: &mut Xoshiro256Plus) {
    let Script::Commands(cmds) = script;
    for cmd in cmds {
        fuzz_cmd(&mut cmd.borrow_mut(), rng);
    }
}

fn fuzz_cmd(cmd: &mut Command, rng: &mut Xoshiro256Plus) {
    match cmd {
        Command::Assert(sexp) => fuzz_sexp(&mut sexp.borrow_mut(), rng),
        _ => (),
    }
}

fn fuzz_sexp(sexp: &mut SExp, rng: &mut Xoshiro256Plus) {
    match sexp {
        SExp::Constant(c) => fuzz_const(&mut c.borrow_mut(), rng),
        SExp::Compound(v)
        | SExp::BExp(_, v)
        | SExp::NExp(_, v)
        | SExp::FPExp(_, _, v)
        | SExp::StrExp(_, v) => {
            for rec_sexp in v {
                fuzz_sexp(&mut rec_sexp.borrow_mut(), rng)
            }
        }
        SExp::QExists(_, rec_sexp) | SExp::QForAll(_, rec_sexp) => {
            fuzz_sexp(&mut *rec_sexp.borrow_mut(), rng)
        }
        SExp::Let(bindings, rec_sexp) => {
            for (_, binding_sexp) in bindings {
                fuzz_sexp(&mut binding_sexp.borrow_mut(), rng)
            }
            fuzz_sexp(&mut rec_sexp.borrow_mut(), rng)
        }
        SExp::Symbol(_) => (),
    }
}

struct BinStr {
    buf: Vec<u8>,
    chars: usize,
}

impl BinStr {
    fn new_random(num_chars: usize, rng: &mut Xoshiro256Plus) -> Self {
        let buf_len = num_chars / 8 + 1;
        let mut v = vec![0; buf_len];
        rng.fill(&mut v[..]);
        BinStr {
            buf: v,
            chars: num_chars,
        }
    }
}

impl fmt::Display for BinStr {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for (i, z) in self.buf.iter().enumerate() {
            if 8 * (i + 1) > self.chars {
                for j in 8 * i..self.chars {
                    write!(
                        f,
                        "{}",
                        format!("{:b}", z)
                            .chars()
                            .nth(j - 8 * i)
                            .unwrap()
                            .to_string()
                    )?;
                }
            } else if 8 * i < self.chars {
                write!(f, "{:08b}", z)?;
            } else {
                break;
            }
        }
        Ok(())
    }
}

fn fuzz_const(constant: &mut Constant, rng: &mut Xoshiro256Plus) {
    let b: &mut [u8] = &mut [0; 4];
    rng.fill(b);
    let mut u = Unstructured::new(b);

    match constant {
        Constant::UInt(_) => {
            *constant = Constant::UInt(u64::arbitrary(&mut u).unwrap().to_string())
        }
        Constant::Dec(_) => *constant = Constant::Dec(f64::arbitrary(&mut u).unwrap().to_string()),
        Constant::Str(_) => *constant = Constant::Str(String::arbitrary(&mut u).unwrap()),
        Constant::Bool(_) => *constant = Constant::Bool(rng.gen_bool(0.5)),
        Constant::Bin(_) | Constant::Hex(_) | Constant::Fp(_) => match constant.sort() {
            Sort::BitVec(length) => {
                *constant = Constant::Bin(
                    format!("{}", BinStr::new_random(length as usize, rng))
                        .chars()
                        .collect(),
                )
            }
            Sort::Fp(e, s) => {
                let e = e.parse::<usize>().unwrap();
                let s = s.parse::<usize>().unwrap();
                let signb = if rng.gen_bool(0.5) {
                    BitVecConst::Bin(vec!['1'])
                } else {
                    BitVecConst::Bin(vec!['0'])
                };
                let eb =
                    BitVecConst::Bin(format!("{}", BinStr::new_random(e, rng)).chars().collect());
                let sb = BitVecConst::Bin(
                    format!("{}", BinStr::new_random(s - 1, rng))
                        .chars()
                        .collect(),
                );

                *constant = Constant::Fp(FPConst::Num(signb, eb, sb))
            }
            _ => panic!("Unreachable Code!"),
        },
    }
}

mod test {
    use super::*;

    use crate::parser::script;
    use insta::assert_display_snapshot;

    #[test]
    fn randomized_baseline() {
        let mut script =
            script("(declare-fun z () Real)(assert (= \" \" #xfff (+ #b101 4 (_ +zero 2 4))))")
                .unwrap()
                .1;
        let mut md = Metadata::new_empty();
        let mut rng = Xoshiro256Plus::seed_from_u64(1);
        fuzz(&mut script, &mut rng);
        assert_display_snapshot!(script);
    }
}
