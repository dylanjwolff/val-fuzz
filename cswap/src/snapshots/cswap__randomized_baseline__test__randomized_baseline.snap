---
source: cswap/src/randomized_baseline.rs
expression: script
---
(assert (= -24076450876586747000000000 "o"))
(assert (= #b101000001101 (+ #b010 493062418185688632 (fp #b1 #b00 #b000))))

